### PR TITLE
feat(api): Get Platform Topology

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description
GET endpoint /platform/topology to retrieve a topology with a JSON Graph Schema Responsse

**Fixes** # MON-5720

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.04.x
- [ ] 19.10.x
- [ ] 20.04.x
- [x] 20.10.x 
- [x] 21.04.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

## Checklist

- [x] I followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Comment by Nohzoh** [comment]
We should avoid references to Helios in the core code.

- This endpoint provides the platform topology in a jsongraph format. There is no need to precise that it is for Helios.
- It might be used for purpose other than Helios.
- And Helios is the name of a service that could be changed further.

In my opinion, this is a bad idea to use the name "Helios" in the whole code of this PR.

File: tests/php/Centreon/Domain/PlatformTopology/PlatformTopologyServiceTest.php [hunk 1]
```diff
@@ -282,9 +294,201 @@ public function testAddPlatformToTopologySuccess(): void
...
+            $this->monitoringServerService,
+            $this->brokerRepository,
+            $this->platformTopologyRegisterRepository
+        );
+
+        /**
+         * Poller Case
+         */
+        $this->expectException(EntityNotFoundException::class);
+        $this->expectExceptionMessage("Topology address for parent platform ID: '3' not found");
```
Comment (sc979) [inline tests/php/Centreon/Domain/PlatformTopology/PlatformTopologyServiceTest.php:410]: IMHO : no exceptions should be returned when unconsistent data are found.

File: src/Centreon/Application/Controller/PlatformTopologyController.php [hunk 2]
```diff
@@ -123,10 +128,52 @@ public function addPlatformToTopology(Request $request): View
...
+     *
+     * @return View
+     */
+    public function getPlatformJsonGraph(): View
+    {
+        $this->denyAccessUnlessGrantedForApiConfiguration();
+
+        try {
+            // Get the entire topology of the platform as an array of PlatformTopology instances
+            $platformTopology = $this->platformTopologyService->getPlatformTopology();
```
Comment (sc979) [inline src/Centreon/Application/Controller/PlatformTopologyController.php:149]:
```suggestion
            /*
             * @var Platform[] $platformTopology
             */
            $platformTopology = $this->platformTopologyService->getPlatformTopology();
```

File: src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php [hunk 3]
```diff
@@ -524,31 +542,83 @@ private function findParentPlatform(PlatformTopology $platformTopology): ?Platfo
...
+    /**
+     * @inheritDoc
+     */
+    public function getPlatformTopology(): array
+    {
+        // find all entries of the platform topology
+        $PlatformTopology = $this->platformTopologyRepository->getPlatformTopology();
+        if ($PlatformTopology === null) {
+            throw new EntityNotFoundException('Platform Topology not found');
+        }
```
Comment (sc979) [inline src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php:582]:
```suggestion
        if ($PlatformTopology === null) {
            throw new EntityNotFoundException(_('No Platform Topology found'));
        }
```

Comment (sc979) [reply src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php:582]: Kindly add the french translation too

File: src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php [hunk 4]
```diff
@@ -524,31 +542,83 @@ private function findParentPlatform(PlatformTopology $platformTopology): ?Platfo
...
         return null;
     }
+
+    /**
+     * @inheritDoc
+     */
+    public function getPlatformTopology(): array
+    {
+        // find all entries of the platform topology
+        $PlatformTopology = $this->platformTopologyRepository->getPlatformTopology();
```
Comment (sc979) [inline src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php:579]: $PlatformTopology -> $platformTopology

File: src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php [hunk 5]
```diff
@@ -524,31 +542,83 @@ private function findParentPlatform(PlatformTopology $platformTopology): ?Platfo
...
+        if ($PlatformTopology === null) {
+            throw new EntityNotFoundException('Platform Topology not found');
+        }
+
+        foreach ($PlatformTopology as $platform) {
+            // Check if the parent are correctly set
+            if ($platform->getType() !== Platform::TYPE_CENTRAL) {
```
Comment (sc979) [inline src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php:586]: kindly check with the PM for the remote top level case

File: src/Centreon/Domain/PlatformTopology/PlatformRelation.php [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    /**
+     * Set the value of relation
+     *
+     * @param  string  $relation
+     *
+     * @return  self
+     */
+    public function setRelation(string $relation): self
```
Comment (sc979) [inline src/Centreon/Domain/PlatformTopology/PlatformRelation.php:90]: JAC : As discussed (https://github.com/centreon/centreon/pull/9128/files#diff-f6f9666fe2d8e90852d098c972983d6314c938d10c2189a1cf44b612da7ecae5R612)

**Review by sc979 — CHANGES_REQUESTED** [review]

File: src/Centreon/Domain/PlatformTopology/Interfaces/PlatformTopologyRepositoryInterface.php [hunk 7]
```diff
@@ -84,7 +84,7 @@ public function getPlatformTopology(): ?array;
      * Search for the address of a topology using its Id
      *
      * @param integer $serverId
-     * @return string|null
+     * @return Platform|null
      */
-    public function findPlatformAddressById(int $serverId): ?string;
+    public function findPlatform(int $serverId): ?Platform;
```
Comment (sc979) [inline src/Centreon/Domain/PlatformTopology/Interfaces/PlatformTopologyRepositoryInterface.php:89]: findPlatform -> findPlatformById

File: src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php [hunk 8]
```diff
@@ -576,23 +576,18 @@ private function findParentPlatform(Platform $platform): ?Platform
...
+        if ($platformTopology === null) {
+            throw new EntityNotFoundException(_('No Platform Topology found'));
         }
 
-        foreach ($PlatformTopology as $platform) {
+        foreach ($platformTopology as $platform) {
             // Check if the parent are correctly set
             if ($platform->getType() !== Platform::TYPE_CENTRAL) {
-                $platformParentAddress = $this->platformTopologyRepository->findPlatformAddressById(
+                $platformParent = $this->platformTopologyRepository->findPlatform(
```
Comment (sc979) [inline src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php:589]: findPlatform -> findPlatformById

File: src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php [hunk 9]
```diff
@@ -576,23 +576,18 @@ private function findParentPlatform(Platform $platform): ?Platform
...
                     $platform->getParentId()
                 );
-                if ($platformParentAddress === null) {
-                    throw new EntityNotFoundException(
-                        sprintf(_("Topology address for parent platform ID: '%d' not found"), $platform->getParentId())
-                    );
-                }
-                $platform->setParentAddress($platformParentAddress);
+                $platform->setParentAddress($platformParent->getAddress());
```
Comment (sc979) [inline src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php:592]: what will happen if the endpoint is called from a remote ?
eg: platformParentAddress === null && $platform->getType !== Platform::TYPE_CENTRAL

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRepositoryRDB.php [hunk 10]
```diff
@@ -178,28 +178,32 @@ public function findLocalMonitoringIdFromName(string $serverName): ?Platform
...
-            if (!empty($platformTopology)) {
-                return $platformTopology;
-            }
         }
+
         return $platformTopology;
     }
 
-    public function findPlatformAddressById(int $serverId): ?string
+    public function findPlatform(int $serverId): ?Platform
```
Comment (sc979) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRepositoryRDB.php:199]: findPlatform -> findPlatformById

File: tests/php/Centreon/Domain/PlatformTopology/PlatformTopologyServiceTest.php [hunk 11]
```diff
@@ -428,12 +395,8 @@ public function testGetPlatformTopologyRelationSetting(): void
 
         $this->platformTopologyRepository
             ->expects($this->exactly(2))
-            ->method('findPlatformAddressById')
-            ->willReturn('1.1.1.1');
-
-        $brokerConfiguration = (new BrokerConfiguration())
-            ->setConfigurationKey('one_peer_retention_mode')
-            ->setConfigurationValue('no');
+            ->method('findPlatform')
```
Comment (sc979) [inline tests/php/Centreon/Domain/PlatformTopology/PlatformTopologyServiceTest.php:398]: findPlatform -> findPlatformById

**Review by sc979 — CHANGES_REQUESTED** [review]

File: src/Centreon/Domain/PlatformTopology/Platform.php [hunk 12]
```diff
@@ -340,10 +340,6 @@ public function getRelation(): ?PlatformRelation
      */
     public function setRelation(?string $relationType): self
     {
-        if (!in_array($relationType, self::AVAILABLE_RELATIONS)) {
```
Comment (sc979) [inline src/Centreon/Domain/PlatformTopology/Platform.php]: Shouldn't this check be added in the relation's setter entity ?
https://github.com/centreon/centreon/blob/eb7172b987df2e942931007a2f93b6c67ab0ce24/src/Centreon/Domain/PlatformTopology/PlatformRelation.php#L93

File: src/Centreon/Domain/PlatformTopology/PlatformRelation.php [hunk 13]
```diff
@@ -92,6 +106,9 @@ public function getRelation(): string
      */
     public function setRelation(string $relation): self
     {
+        if (!in_array($relation, self::AVAILABLE_RELATIONS)) {
+            throw new \InvalidArgumentException(sprintf(_("The type of relation '%s' is not allowed"), $relation));
```
Comment (sc979) [inline src/Centreon/Domain/PlatformTopology/PlatformRelation.php:110]: IMHO : You shouldn't throw an exception here, and instead set a default value

**Review by sc979 — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Pull request merged**